signer/src/api/new_block.rs

```diff
@@ -32,39 +26,119 @@ pub type StacksNodePayload = Result<Json<serde_json::Value>, JsonRejection>;
 /// second might succeed, we will return a 200 OK status code.
 ///
 /// [^1]: <https://github.com/stacks-network/stacks-core/blob/09c4b066e25104be8b066e8f7530ff0c6df4ccd5/testnet/stacks-node/src/event_dispatcher.rs#L317-L385>
-pub async fn new_block_handler<S>(_state: State<S>, body: StacksNodePayload) -> StatusCode
+pub async fn new_block_handler<S>(state: State<ApiState<S>>, body: String) -> StatusCode
 where
     S: DbWrite,
 {
     tracing::info!("Received a new block event from stacks-core");
+    let api = state.0;
+    let network = api.settings.signer.network;
 
-    let _event: NewBlockEvent = match body {
-        Ok(Json(value)) => {
-            // We print out the raw string if there are any errors during
-            // deserialization, but only for non-release builds.
-            #[cfg(debug_assertions)]
-            let raw_payload = serde_json::to_string(&value).unwrap();
-
-            match serde_json::from_value(value) {
-                Ok(event) => event,
-                Err(error) => {
-                    tracing::error!("could not deserialize POST /new_block webhook: {error}");
-
-                    #[cfg(debug_assertions)]
-                    println!("raw payload: {raw_payload}");
-                    return StatusCode::OK;
-                }
-            }
-        }
+    let new_block_event: NewBlockEvent = match serde_json::from_str(&body) {
+        Ok(value) => value,
         // If we are here, then we failed to deserialize the webhook body
         // into the expected type. It's unlikely that retying this webhook
         // will lead to success, so we log the error and return `200 OK` so
         // that the node does not retry the webhook.
         Err(error) => {
-            tracing::error!("could not deserialize POST /new_block webhook: {error:?}");
+            tracing::error!("could not deserialize POST /new_block webhook: {body}, error {error}");
```

Comment:
I think this is fine for now since realistically this shouldn't happen (unless Stacks makes a release that completely borks the json schema... totally possible :sweat_smile: ), but in reality this is probably a _fatal_ situation, isn't it? If we're missing blocks then we're potentially missing important sbtc contract calls?

Reply:
Yeah that makes sense, this shouldn't happen. Although it's more likely that we have some "bug" in our expectation of the schema. If we hit this we'd have to have the stacks node replay the events somehow (no idea how to actually do that).

Reply:
Something in me feels like there maybe should be some p2p chatter around sbtc events, so a signer (and its peers) can determine if it's lagging/faulty... which should probably result in the signer being ejected from the active set?

Reply:
> This isn't possible today :( The events are only emitted as a part of block processing. 

That makes sense. Is there a way to have a node forget/delete blocks past a certain height (this is closer to what I had in mind)? If not we'd have to wipe it clean? Can we do that in a "nice" way?

> Something in me feels like there maybe should be some p2p chatter around sbtc events, so a signer (and its peers) can determine if it's lagging/faulty... which should probably result in the signer being ejected from the active set?

The chatter idea might be a good one 🤔. But they probably shouldn't be ejected if they are falling behind, we've just been designing the signers with that possibility in mind.

Reply:
> That makes sense. Is there a way to have a node forget/delete blocks after a certain height (this is closer to what I had in mind)? If not we'd have to wipe it clean? Can we do that in a "nice" way?

Unfortunately, no :frowning_face:  The concept of "deleting" data is a bit anti-blockchain, so I can understand why rolling-back a node hasn't been part of the design considerations. I'm sure it'd be possible to implement, but I'm not so sure stacks-core would go along with it.

>The chatter idea might be a good one 🤔. But they probably shouldn't be ejected if they are falling behind, we've just been designing the signers with that possibility in mind.

Are there some design docs I could dig into regarding this design so I can get a better idea how it's handled?

Reply:
I'll go ahead and give an approve so you don't get blocked on this discussion, but maybe we should create an issue to track this?

Reply:
> I'll go ahead and give an approve so you don't get blocked on this discussion, but maybe we should create an issue to track this?

Yeah, good point, I'll open one for adding metrics and alerts. That's how I would handle this case.

Reply:
I opened https://github.com/stacks-network/sbtc/issues/481, but it needs some refining.

Reply:
> I'll go ahead and give an approve so you don't get blocked on this discussion, but maybe we should create an issue to track this?

There are a few. For this particular question I think the designs are laid out in https://github.com/stacks-network/sbtc/issues/39 and https://github.com/stacks-network/sbtc/issues/44. There are proper google docs too, with their links in different issues. I generally find them using this query https://github.com/stacks-network/sbtc/issues?q=is%3Aissue+design+is%3Aclosed+LLD, their usually the tickets with LLD in the name.